pkg/start/start.go

```diff
@@ -0,0 +1,316 @@
+	//   time we can remove our changes to OnStartedLeading.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ch := make(chan os.Signal, 1)
+	defer func() { signal.Stop(ch) }()
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		sig := <-ch
+		glog.Infof("Shutting down due to %s", sig)
+		cancel()
+
+		// exit after 2s no matter what
+		select {
+		case <-time.After(2 * time.Second):
+			glog.Fatalf("Exiting")
+		case <-ch:
+			glog.Fatalf("Received shutdown signal twice, exiting")
+		}
+	}()
+
+	o.run(ctx, controllerCtx, lock)
+	return nil
+}
+
+func (o *Options) run(ctx context.Context, controllerCtx *Context, lock *resourcelock.ConfigMapLock) {
+	// listen on metrics
+	if len(o.ListenAddr) > 0 {
+		mux := http.NewServeMux()
+		mux.Handle("/metrics", promhttp.Handler())
+		go func() {
+			if err := http.ListenAndServe(o.ListenAddr, mux); err != nil {
+				glog.Fatalf("Unable to start metrics server: %v", err)
+			}
+		}()
+	}
+
+	exit := make(chan struct{})
+
+	// TODO: when we switch to graceful lock shutdown, this can be
+	// moved back inside RunOrDie
+	go leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          lock,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(stop <-chan struct{}) {
+				controllerCtx.Start(ctx.Done())
+				select {
+				case <-ctx.Done():
+					// WARNING: this is not completely safe until we have Kube 1.14 and ReleaseOnCancel
+					//   and client-go ContextCancelable, which allows us to block new API requests before
+					//   we step down. However, the CVO isn't that sensitive to races and can tolerate
+					//   brief overlap.
+					glog.Infof("Stepping down as leader")
+					// give the controllers some time to shut down
+					time.Sleep(100 * time.Millisecond)
+					// if we still hold the leader lease, clear the owner identity (other lease watchers
+					// still have to wait for expiration) like the new ReleaseOnCancel code will do.
+					if err := lock.Update(resourcelock.LeaderElectionRecord{}); err == nil {
+						// if we successfully clear the owner identity, we can safely delete the record
+						if err := lock.Client.ConfigMaps(lock.ConfigMapMeta.Namespace).Delete(lock.ConfigMapMeta.Name, nil); err != nil {
+							glog.Warningf("Unable to step down cleanly: %v", err)
+						}
+					}
+					glog.Infof("Finished shutdown")
+					close(exit)
+				case <-stop:
+					// we will exit in OnStoppedLeading
+				}
+			},
+			OnStoppedLeading: func() {
+				glog.Warning("leaderelection lost")
+				close(exit)
+			},
+		},
+	})
+
+	<-exit
+}
+
+// createResourceLock initializes the lock.
+func createResourceLock(cb *ClientBuilder, namespace, name string) (*resourcelock.ConfigMapLock, error) {
+	recorder := record.NewBroadcaster().NewRecorder(runtime.NewScheme(), v1.EventSource{Component: namespace})
```

Comment:
ques: shouldn't `runtime.NewScheme()` be `kubernetes.Scheme` ?

Reply:
I’m not 100% sure, it doesn’t actually matter for events (we fall back to a generic path if an object isn’t in the scheme)

Reply:
Actually, we didn't have a sink, so we weren't actually recording metrics.  I fixed that here, and made sure each of the places in the CVO we log events we were sending it to our namespace rather than the default namespace (which if you pass "" is where you go).  Added an integration test to verify we receive a leader event, which is a good proxy for us accidentally breaking it in the future.